src/cargo/util/toml/mod.rs

```diff
@@ -610,10 +645,18 @@ fn resolve_dependencies<'a>(
             }
         }
 
-        deps.insert(
-            name_in_toml.clone(),
-            manifest::InheritableDependency::Value(resolved.clone()),
-        );
+        // if the dependency is not optional, it is always used
+        // if the dependency is optional and activated, it is used
+        // if the dependency is optional and not activated, it is not used
+        let is_dep_activated =
+            !resolved.is_optional() || activated_opt_deps.contains(name_in_toml.as_str());
+        // If the edition is less than 2024, we don't need to check for unused optional dependencies
+        if edition < Edition::Edition2024 || is_dep_activated {
```

Comment:
We should have a `cargo publish` test to make sure the published `Cargo.toml` and the `Summary` has these stripped

Reply:
Do we also need a test that shows it's not in the lockfile?